src/python_bindings/mod.rs

```diff
@@ -1,98 +1,165 @@
-use crate::index::{FSMInfo, Index};
+use std::sync::Arc;
+
+use crate::index::Index;
 use crate::json_schema;
 use crate::prelude::*;
-use crate::regex::get_token_transition_keys;
-use crate::regex::get_vocabulary_transition_keys;
-use crate::regex::state_scan_tokens;
-use crate::regex::walk_fsm;
 use bincode::config;
+use bincode::{Decode, Encode};
 use pyo3::exceptions::PyValueError;
 use pyo3::prelude::*;
-use pyo3::types::PyDict;
+use pyo3::types::{PyAny, PyDict};
 use pyo3::wrap_pyfunction;
-use rustc_hash::{FxHashMap, FxHashSet};
+use rustc_hash::{FxHashMap as HashMap, FxHashSet as HashSet};
 use serde_json::Value;
+use tokenizers::FromPretrainedParameters;
+
+macro_rules! type_name {
+    ($obj:expr) => {
+        // Safety: obj is always initialized and tp_name is a C-string
+        unsafe { std::ffi::CStr::from_ptr((&*(&*$obj.as_ptr()).ob_type).tp_name) }
+    };
+}
```

Comment:
This looks very fragile to me. Are there any alternatives to this?

PyO3 can call python right, maybe we can get this info by calling python?

Reply:
It creates a C string from a pointer and to get to that pointer it dereferences other raw pointers. It's probably okay as indicated by the tests, but still if there is a way to do this without using unsafe, it'd be better IMO.

Reply:
Not all unsafe is unsafe, safety here is provided by the fact that these objects are always initialized, if `ob_type` won't be then whole python will be broken at the core, the same goes for `tp_name`: https://docs.python.org/3/c-api/typeobj.html#c.PyTypeObject.tp_name:

> - This field must not be NULL
> - `OT` markers in the table

Weirdly I didn't find a way to get an object's name via public interface of pyo3.

And going around back to python definitely is not justified for this case, where we just trying be nice and to provide helpful messages to the user. Especially, when there is a quite safe way to get this information.